### PR TITLE
Feature: pip support

### PR DESCRIPTION
**Pull request by mattkram**
Adds support for `pip` dependencies. For this, we take the same approach as `conda-lock`.

In addition, I have modified the logic of `Environment.is_prepared` to do a more robust comparison of packages installed vs. specified in the lock file.

### CODE REVIEW

**Comment by AlbertDeFusco** [comment]
Thanks. I guess I should pull the windows fix to a separate PR

**Comment by mattkram** [comment]
Yeah I was going to do that after I saw these results 🙂.

Could use feedback on whether it's okay to check return code or if there's
a better way to check if a package got installed via pip.

-Matt

Sent via Mobile

**Comment by AlbertDeFusco** [comment]
I would look into using conda list, which includes pip packages to verify that they were installed. This could be a small change to is_prepared to include a check for any pip packages, which would also be good to do even if no pip packages appear in the env sources. 

I think the pip install return code will be ok as well.

File: conda_project/project.py [hunk 1]
```diff
@@ -111,25 +113,38 @@ def is_prepared(self) -> bool:
...
-        return False
+        # Generate a set of (name, version, manager) tuples from the conda environment
+        # We also convert the conda package_type attribute to a string in the set
+        # {"conda", "pip"} to allow direct comparison with conda-lock.
+        # TODO: Consider comparing more than the name, version, & manager
+        # TODO: pip_interop_enabled is marked "DO NOT USE". What is the alternative?
+        manager_map = {PackageType.VIRTUAL_PYTHON_WHEEL: "pip"}
+        pd = PrefixData(self.prefix, pip_interop_enabled=True)
+        installed_pkgs = {
+            (p.name, p.version, manager_map.get(p.package_type, default="conda"))
```
Comment (mattkram) [inline conda_project/project.py:130]:
```suggestion
            (p.name, p.version, manager_map.get(p.package_type, "conda"))
```

**Comment by AlbertDeFusco** [comment]
Cool. I'll take a look. I've been working on this, too and was using sha256 for pip and conda

https://github.com/AlbertDeFusco/conda-project/commit/4c44dfe6bea3662dd0164cfa7ae499f26e59fc6a

**Comment by AlbertDeFusco** [comment]
wow, good catch on the prefixdata cache. That's why my tests were failing.

**Comment by mattkram** [comment]
That was a tricky one!

I think the general skeleton is there in this PR, if sha256 would work better feel free to swap that out.

**Comment by AlbertDeFusco** [comment]
You're approach looks much more streamlined than I did. I bet we can just as easily use sha256. 

What I found with my tests was some unfortunate slowdowns in the is_prepared property. I'll run yours a few times and see how it does. 

Either way, I think we'll need to see about further optimizations here (and maybe even ensuring the is_prepared is called as infrequently as possibler

**Comment by mattkram** [comment]
I can take a look at yours and pull out the relevant bits. I saw you had more extensive handling of the different types of pip packages (I only found one).

I agree a property may be inefficient here. Probably something to sort out in another PR. I actually had the realization that this whole thing is a finite state machine so actually we may be able to simplify all the methods a lot with some refactoring. 

For now, I propose I modify this PR to capture all package types associated with pip packages and compare based on sha256 hash instead of package name and version. I'll work on that this evening probably.

**Comment by AlbertDeFusco** [comment]
That would be great to explore. Right now I've been going after an exhaustive comparison between source spec, lock, and live env. Perhaps there are faster ways to determine that "something" doesn't mathch rather then working through all packages. I imagine it can only get worse for large envs

File: conda_project/project.py [hunk 2]
```diff
@@ -17,6 +17,8 @@
 from subprocess import SubprocessError
 from typing import List, Optional, Tuple, Union
 
+from conda_lock._vendor.conda.core.prefix_data import PrefixData
```
Comment (AlbertDeFusco) [inline conda_project/project.py:20]: One word of caution. In Conda Lock 1.2 it is `conda_lock._vendor...`, but earlier versions use `conda_lock.vendor...`

Comment (mattkram) [reply conda_project/project.py:20]: I was wondering why my local version was broken compared to CI, and then got fixed when I did `conda update`  ... (see [6b3ea42](https://github.com/conda-incubator/conda-project/pull/39/commits/6b3ea42a066831722a449266151ee896df4c237c))

We should probably pin to >1.2 unless we want to do a conditional import ...

**Comment by mattkram** [comment]
Alrighty, made a few cleanups and added sha256 to the comparison. I have left in package name, version, and manager since they should be lightweight and I figured it opens the possibility to have better messaging for why an environment may not match the lockfile. Let me know if you think I should remove.

**Comment by mattkram** [comment]
Okay, I think I've got it where it needs to be. I think the tests should pass now. Let's see ..

File: conda_project/project.py (hunk 2 above)
Comment (AlbertDeFusco) [reply conda_project/project.py:21]: Pinning to 1.2 sounds ok to me.

**Review by AlbertDeFusco — APPROVED** [review]

**Comment by AlbertDeFusco** [comment]
Can you also edit the setup.py with the pin

**Pull request merged**